File: lib/src/main/java/com/walmartlabs/components/scheduler/core/CheckpointHelper.java
package com.walmartlabs.components.scheduler.core;

import com.fasterxml.jackson.core.type.TypeReference;
import com.google.common.base.Function;
import com.google.common.base.Joiner;
import com.google.common.base.Splitter;
import com.google.common.collect.ImmutableMap;
import com.google.common.util.concurrent.ListenableFuture;
import com.walmart.gmp.ingestion.platform.framework.data.core.DataManager;
import com.walmartlabs.components.scheduler.entities.EventLookup;
import com.walmartlabs.components.scheduler.entities.EventLookupDO.EventLookupKey;
import org.apache.log4j.Logger;

import java.io.IOException;
import java.time.ZonedDateTime;
import java.util.*;

import static com.google.common.collect.Maps.transformValues;
import static com.google.common.util.concurrent.Futures.*;
import static com.walmart.gmp.ingestion.platform.framework.data.core.Selector.fullSelector;
import static com.walmart.gmp.ingestion.platform.framework.messaging.kafka.Constants.OBJECT_MAPPER;
import static java.lang.Integer.parseInt;

/**
 * Created by smalik3 on 7/5/16
 */
public class CheckpointHelper {

    private static final Logger L = Logger.getLogger(CheckpointHelper.class);

    private static final ListenableFuture<EventLookup> SUCCESS = immediateFuture(null);

    private final DataManager<EventLookupKey, EventLookup> lookupDataManager;
    private final EventLookupKey KEY = new EventLookupKey("_CHECKPOINT_", "_CHECKPOINT_");

    @SuppressWarnings("unchecked")
    CheckpointHelper(BucketManager bucketManager) {
        this.lookupDataManager = (DataManager) bucketManager.getEventDataManager();
    }

    synchronized ListenableFuture<EventLookup> saveCheckpoint(Map<ZonedDateTime, BucketSnapshot> data) {
        try {
            final Map<ZonedDateTime, BucketSnapshot> sorted = new TreeMap<>(data);
            if (L.isDebugEnabled())
                L.debug("saving checkpoint for buckets: " + sorted.keySet());
            final EventLookup entity = DataManager.entity(EventLookup.class, KEY);
            entity.setPayload(OBJECT_MAPPER.writeValueAsString(transformValues(sorted, (Function<BucketSnapshot, Map<String, Object>>) bs ->
                    ImmutableMap.of("b", bs.id().toString(), "c", bs.count(), "a", from(bs.awaiting()), "p", from(bs.processing()))).values()));
            return lookupDataManager.saveAsync(entity);
        } catch (Exception e) {
            L.error("error in saving the checkpoint: buckets: " + data, e);
            return SUCCESS; // we do not want to suppress further executions of this timer
        }
    }

    ListenableFuture<Map<ZonedDateTime, BucketSnapshot>> loadCheckpoint() {
        final ListenableFuture<EventLookup> f = lookupDataManager.getAsync(KEY, fullSelector(KEY));
        return catching(transform(f, (Function<EventLookup, Map<ZonedDateTime, BucketSnapshot>>) result -> {
            try {
                if (result == null) {
                    L.info("no checkpoint to load");
                    return new HashMap<>();
                } else if (result.getPayload() != null && result.getPayload().trim().length() > 0) {
                    final List<Map<String, Object>> l = OBJECT_MAPPER.readValue(result.getPayload(), new TypeReference<List<Map<String, Object>>>() {
                    });
                    final Map<ZonedDateTime, BucketSnapshot> data = new HashMap<>();
                    l.forEach(d -> {
                        final ZonedDateTime id = ZonedDateTime.parse(d.get("b").toString());
                        final BitSet awaiting = to(d.get("a").toString());
                        final BitSet processing = to(d.get("p").toString());
                        processing.stream().forEach(awaiting::set);
                        data.put(id, new BucketSnapshot(id, (Integer) d.get("c"), awaiting, new BitSet()));
                    });
                    return data;
                } else return new HashMap<>();
            } catch (IOException e) {
                L.error("error in loading the checkpoint, ignoring", e);
                return new HashMap<>();
            }
        }), Exception.class, ex -> {
            L.error("error in loading checkpoint, ignoring", ex);
            return new HashMap<>();
        });
    }

    private String from(BitSet bitSet) {
        return Joiner.on(",").join(bitSet.stream().boxed().toArray());
    }

    private BitSet to(String bits) {
        final BitSet bs = new BitSet();
        if (bits == null || bits.trim().length() == 0)
            return bs;
        Splitter.on(",").split(bits).forEach(s -> bs.set(parseInt(s)));
        return bs;
    }

}

File: lib/src/main/kotlin/com/walmartlabs/opensource/bigben/core/CheckpointHelper.kt
package com.walmartlabs.opensource.bigben.core

import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.annotation.JsonInclude.Include.NON_EMPTY
import com.google.common.util.concurrent.ListenableFuture
import com.walmartlabs.opensource.bigben.entities.EventLookup
import com.walmartlabs.opensource.bigben.extns.fetch
import com.walmartlabs.opensource.bigben.extns.save
import com.walmartlabs.opensource.core.*
import java.time.ZonedDateTime
import java.util.*
import java.util.stream.Collectors

internal class CheckpointHelper {
    companion object {
        private val l = logger<CheckpointHelper>()
        private val CHECKPOINT_KEY = "_CHECKPOINT_"
    }

    @JsonInclude(NON_EMPTY)
    data class Checkpoint(val b: String, val c: Long, val a: List<Int> = emptyList(), val p: List<Int> = emptyList())

    fun saveCheckpoint(data: Map<ZonedDateTime, BucketSnapshot>): ListenableFuture<EventLookup> {
        return data.toSortedMap().mapValues {
            Checkpoint(it.value.id.toString(), it.value.count,
                    it.value.awaiting.stream().boxed().collect(Collectors.toList()),
                    it.value.processing.stream().boxed().collect(Collectors.toList()))
        }.let { m ->
            if (l.isDebugEnabled) l.debug("saving checkpoint for buckets: {}", m.keys)
            save<EventLookup> { it.tenant = CHECKPOINT_KEY; it.xrefId = CHECKPOINT_KEY; it.payload = m.values.json() }
        }.done({ l.warn("error in saving checkpoint", it.rootCause()) }) {
            if (l.isDebugEnabled && it != null) l.debug("checkpoint saved successfully for buckets: {}", it.payload)
        }
    }

    fun loadCheckpoint(): ListenableFuture<Map<ZonedDateTime, BucketSnapshot>> {
        return fetch<EventLookup> { it.tenant = CHECKPOINT_KEY; it.xrefId = CHECKPOINT_KEY }.transform {
            when {
                it?.payload != null && it.payload!!.trim().isNotEmpty() -> {
                    typeRefJson<List<Checkpoint>>(it.payload!!).map {
                        val awaiting = it.a.fold(BitSet(), { b, i -> b.apply { set(i) } })
                        val processing = it.p.fold(BitSet(), { b, i -> b.apply { set(i) } })
                        processing.stream().forEach { awaiting.set(it) }
                        BucketSnapshot(ZonedDateTime.parse(it.b), it.c, BitSet(), awaiting)
                    }.associate { it.id to it }.also { if (l.isDebugEnabled) l.debug("loaded checkpoint: {}", it) }
                }
                else -> {
                    if (l.isInfoEnabled)
                        l.info("no checkpoint to load")
                    HashMap()
                }
            }
        }.catching { l.error("error in loading checkpoint, ignoring", it.rootCause()); HashMap() }
    }
}